Add scantxoutset RPC method

jonasschnelli opened the pull request:
Alternative to #9152.

This takes `<n>` pubkeys and optionally  `<n>` xpubs (together with a definable lookup windows where the default is 0-1000) and looks up common scripts in the UTXO set of all given or derived keys.

The output will be an array similar to `listunspent`. That array is compatible with `createrawtransaction` as well as with `signrawtransaction`.

This makes it possible to prepare sweeps and have them signed in a secure (cold) space.

jonasschnelli committed:
Add FindScriptPubKey() to search the UTXO set

jonasschnelli committed:
Blockchain/RPC: Add scantxoutset method to scan UTXO set

jonasschnelli committed:
[QA] Add scantxoutset test

gmaxwell commented:
Why is it pubkeys and not addresses for the pubkey part?  (obviously xpubs are xpubs and need to be)

jonasschnelli commented:
After a short discussion on IRC (https://botbot.me/freenode/bitcoin-core-dev/2018-01-16/?msg=95804115&page=2) support for addresses and pubkeys makes most sense. Will add support for an array of addresses.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+    else if (request.params[0].get_str() == "abort") {
+        CoinsViewScanReserver reserver;
+        if (reserver.reserve()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "No scan in progress");
```
IMO there is no need to throw, a `bool` in the response is enough? Otherwise, missing test.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+        throw std::runtime_error(
+            "scantxoutset <action> {\"pubkeys\": [\"pubkey\",...], \"xpubs\":[{\"xpub\": \"<xpub>\"}], other options}\n"
```
Remove `other options`? There are none.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+
+UniValue scantxoutset(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
```
`request.params.size() != 2`?

jonasschnelli replied:
Nah. You can also call `scantxoutset status`

promag replied:
Commit "Blockchain/RPC: Add scantxoutset method to scan UTXO set"

> Nah. You can also call scantxoutset status

Right.

However should be `request.params.size() > 3` (instead of 2)?

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+            "      \"pubkeys\":[\"pubkey\",...]   (array of strings, optional) An array of HEX encoded public keys\n"
+            "      \"xpubs\":                   (array of xpub objects that will be used to derive child keys with the given lookup window after m/0/k and m/1/k scheme)\n"
+            "           [\n"
+            "               {\"xpub\":\"<xpub>\", (base58check encoded extended public key (xpub)\n"
```
Nit, newline after `{`.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+    if (request.params[0].get_str() == "status") {
+        CoinsViewScanReserver reserver;
+        if (reserver.reserve()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "No scan in progress");
```
Missing test for error.

jonasschnelli replied:
You mean missing a functional test for that case. Yes. I though about it, but would require a mockup-slowdown argument (`-testslowdown` or similar). Otherwise I guess it's hard to properly test this.

promag replied:
Yes, agree that `status` should not raise an error when no scan is in progress.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+        }
+        result.push_back(Pair("unspents", unspents));
+        result.push_back(Pair("total_amount", ValueFromAmount(total_in)));
+    }
```
Nit, `else {` here.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "No scan in progress");
+        }
+        g_should_abourt_scan = true;
+    }
```
Nit, `else if () {` here.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+static std::mutex g_utxosetscan;
+static std::atomic<int> g_scan_progress;
+static std::atomic<bool> g_scan_in_progress;
+static std::atomic<bool> g_should_abourt_scan;
```
Typo, abort.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+    }
+
+    ~CoinsViewScanReserver() {
+        std::lock_guard<std::mutex> lock(g_utxosetscan);
```
Lock only when changing `g_scan_in_progress`?
```cpp
if (m_could_reserve) {
    std::lock_guard<std::mutex> lock(g_utxosetscan);
    g_scan_in_progress = false;
}
```

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+    else if (request.params[0].get_str() == "start") {
+        CoinsViewScanReserver reserver;
+        if (!reserver.reserve()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Scan already in progress, use action \"abort\" \"status\"");
```
Remove "status"? Otherwise add ` or ` between actions.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+                result.push_back(Pair("searched_items", count));
+            }
+
+            for (auto& it : coins) {
```
cont auto&

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+                unspent.push_back(Pair("scriptPubKey", HexStr(txo.scriptPubKey.begin(), txo.scriptPubKey.end())));
+                unspent.push_back(Pair("amount", ValueFromAmount(txo.nValue)));
+                unspent.push_back(Pair("height", (int32_t)coin.nHeight));
+
```
Nit, remove newline.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1599,236 @@ UniValue savemempool(const JSONRPCRequest& request)
+    explicit CoinsViewScanReserver() : m_could_reserve(false) {}
+
+    bool reserve() {
+        std::lock_guard<std::mutex> lock(g_utxosetscan);
```
Assert not reserved?
```cpp
assert (!m_could_reserve);
std::lock_guard<std::mutex> lock(g_utxosetscan);
if (g_scan_in_progress) return false;
g_scan_in_progress = true;
m_could_reserve = true;
return true;
```

jonasschnelli replied:
Fixed.

laanwj commented:
Concept ACK, nice!

jonasschnelli commented:
* Added support for addresses (can scan unspent outputs after given addresses)
* Added support for an optional raw sweep transaction including optional feerate or optional confirmation target

jonasschnelli commented:
The raw sweep fee calculation is currently WIP (misses the dummy signer part)... will fix soon.

jonasschnelli commented:
Overhauled the fee calculation logic (see the dummy sign keystore).

gmaxwell commented on src/rpc/blockchain.cpp:
```diff
@@ -1597,6 +1602,335 @@ UniValue savemempool(const JSONRPCRequest& request)
+
+        // flush utxo state and start the scan
+        FlushStateToDisk();
+        bool res = pcoinsdbview->FindScriptPubKey(g_scan_progress, g_should_abort_scan, count, needles, coins);
```
What prevents the state from being mutated out from under this scan?

jonasschnelli replied:
I think due to the time required to perform a scan, it's something that may be tolerated (although it should be mentioned in the docs). Not sure, but I guess it's the same with `gettxoutsetinfo`.

Not sure if you can scan a `CCoinsView` of a snapshot state... I guess no.
Locking cs_main would be "meh".

jonasschnelli replied:
Also, the rawtx (sweep) is in the same way "outdated" the moment you have received it... maybe you could argue that this is the same for `fundrawtx`

sipa replied:
@gmaxwell @jonasschnelli The cursor iterates over the state of the CCoinsView at the time it was created; modifying it during iteration is fine. This only works because `GetCursor` is not implemented for `CCoinsViewCache`, and is invoked directly on the `CCoinsViewDB` LevelDB wrapper.

The downside is that this requires a full flush of the database, hurting performance for all of the process (including validation).

luke-jr commented:
NACK supporting addresses. Addresses have no relation to the UTXOs.

jonasschnelli commented:
@luke-jr: Why? Addresses are encoded output scripts (scriptPubKey). The rational behind supporting addresses is that a) it may be more efficient then forming every possible known common script from a pubkey and b) that pubkeys are somewhat more difficult to export then the pure "used addresses".

luke-jr commented:
Addresses are *opaque* identifiers for a given invoice. That they are currently implemented by encoding a scriptPubKey is irrelevant.

greenaddress commented:
Does it support mempool/unconfirmed utxos? I had a quick look and didn't seem to, I think it would be useful to have mempool too.

Would it make sense to avoid the background job? maybe by keeping the utxo set sorted by scriptPubKey and binary search on it or perhaps some utxo set limited indexing?

jonasschnelli commented:
> Does it support mempool/unconfirmed utxos? I had a quick look and didn't seem to, I think it would be useful to have mempool too.

It currently does not scan the mempool (hence the command name `scantxoutset`), but I agree, that would be useful. But, since scans take a while, timing may be a problem for scanning the mempool.
Maybe an additional RPC call would make sense (`scanmempool`)?

> Would it make sense to avoid the background job? maybe by keeping the utxo set sorted by scriptPubKey and binary search on it or perhaps some utxo set limited indexing?

You mean speeding up the scan? I don't think its worth to keep an extra index (changing the sort order would probably slow down verification a lot).
My local tests did show a mainnet scan takes about ~30seconds (SSD, fast CPU).
I don't know what the space requirements for a by-scriptPubKey-index would be, .... maybe we can look into this once this PR did proceed.

jonasschnelli commented:
Rebased

luke-jr commented:
The commit separation here is ugly: CCoinsView::FindScriptPubKey initially checks ShutdownRequested directly, and then this is removed with the RPC changes.

jonasschnelli commented:
Fixed the ugly commit separation in `coins.cpp`

jonasschnelli commented:
Rebased

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1640,249 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            }
+        }
+
+        // add all common scripts for the given and derived pubkeys
```
It's unfortunate that this introduces yet another "default set of addresses derived from a given key". Maintaining this will lead to incompatibilities between implementations as new types of scripts get added.

It would be better to have an explicit way to describe a set of scripts to watch for that could be reused in multiple places. I'm working towards that (see https://gist.github.com/sipa/125cfa1615946d0c3f3eec2ad7f250a2), but it may be a while before it's usable.

However, would you consider changing the RPC interface to be more easily adaptable to such a design? I think an interface which just takes a list of descriptions, each of which is a JSON object with either an address, script, or xpub (where the xpub one also takes a window size in a separate field). That means we can later add extra fields to describe what derivations to use, for example.

sipa reviewed (commented):
GitHub doesn't seem to let me add more comments on specific line numbers, so I'll put it here:

`CPubKey pub_key` should be const, and styled like a constant (all uppercase).

Overall, I'm unsure about this. This is functionality that is more easily provided by software that maintains a UTXO index by script, and is not possible in general if we'd move to a design like UHF (see mailinglist) or other UTXO avoidance techniques. Those are far away of course, and features like this can be made optional (like `txindex` is) if needed. I'm just generally unconvinced a full node is the best place to put this.

MarcoFalke commented on test/functional/rpc_scantxoutset.py:
```diff
@@ -0,0 +1,70 @@
+        self.log.info("Mining blocks...")
+        self.nodes[0].generate(110)
+
+        addr_P2SH_SEGWIT = self.nodes[0].getnewaddress()
```
nit: specify `p2sh-segwit`, so the tests don't break when the defaults change.

MarcoFalke commented on test/functional/rpc_scantxoutset.py:
```diff
@@ -0,0 +1,70 @@
+
+        self.log.info("Stop node, remove wallet, mine again some blocks...")
+        self.stop_node(0)
+        shutil.rmtree(os.path.join(self.options.tmpdir, "node0/regtest/wallets"))
```
nit: `os.path.join(self.nodes[0].datadir, "regtest", 'wallets')`

MarcoFalke commented on test/functional/rpc_scantxoutset.py:
```diff
@@ -0,0 +1,70 @@
+        self.nodes[0].generate(1)
+
+        self.stop_node(0)
+        os.remove(os.path.join(self.options.tmpdir, "node0/regtest/wallet.dat"))
```
nit: `os.path.join(self.nodes[1].datadir, "regtest", 'wallet.dat'`

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1704,6 +1729,9 @@ UniValue scantxoutset(const JSONRPCRequest& request)
             " \"total_amount\" : x.xxx,          (numeric) The total amount of all found unspent outputs in " + CURRENCY_UNIT + "\n"
+            " \"rawsweep_tx\" : \"value\",       (string) The hex-encoded raw transaction of the optional sweep transaction\n"
+            " \"rawsweep_vsize\" : \"value\",     (numeric) virtual transaction size of the sweep transaction including signatures\n"
+            " \"rawsweep_fee\" : \"value\",       (numeric) Estimated fee for the sweep transaction in " + CURRENCY_UNIT + "\n"
```
Nit: "Estimated fee" implies that the fee is unknown, which is not true given that the exact fee can be calculated.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1671,6 +1674,23 @@ class CoinsViewScanReserver
+class CCoinsViewScanDummySignKeyStore : public CBasicKeyStore
+{
+public:
+    bool GetPubKey(const CKeyID &address, CPubKey& vchPubKeyOut) const {
```
can be marked override

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1671,6 +1674,23 @@ class CoinsViewScanReserver
+        vchPubKeyOut = pub_key;
+        return true;
+    }
+    bool GetCScript(const CScriptID &hash, CScript& redeemScriptOut) const {
```
can be marked override

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1704,6 +1729,9 @@ UniValue scantxoutset(const JSONRPCRequest& request)
             "   ,...], \n"
             " \"total_amount\" : x.xxx,          (numeric) The total amount of all found unspent outputs in " + CURRENCY_UNIT + "\n"
+            " \"rawsweep_tx\" : \"value\",       (string) The hex-encoded raw transaction of the optional sweep transaction\n"
+            " \"rawsweep_vsize\" : \"value\",     (numeric) virtual transaction size of the sweep transaction including signatures\n"
```
I think the vsizes of the signatures are only estimated?

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1692,6 +1712,11 @@ UniValue scantxoutset(const JSONRPCRequest& request)
             "           ]\n"
+            "      \"rawsweep\": {\n             (object, optional) Optionally creates a raw sweep transaction\n"
+            "          \"address\": \"address\",   (string, optional) Address where the funds should be sent to\n"
+            "          \"feerate\": n,           (numeric, optional, default not set: makes wallet determine the fee) Set a specific fee rate in " + CURRENCY_UNIT + "/kB\n"
```
I believe this doesn't use the wallet, but the `conf_target` instead?

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1692,6 +1712,11 @@ UniValue scantxoutset(const JSONRPCRequest& request)
+            "      \"rawsweep\": {\n             (object, optional) Optionally creates a raw sweep transaction\n"
+            "          \"address\": \"address\",   (string, optional) Address where the funds should be sent to\n"
+            "          \"feerate\": n,           (numeric, optional, default not set: makes wallet determine the fee) Set a specific fee rate in " + CURRENCY_UNIT + "/kB\n"
+            "          \"conf_target\": n,       (numeric, optional) Confirmation target (in blocks), has no effect if feerate is provided\n"
```
I believe the default for conf_target is 6?

maflcko reviewed (approved):
utACK 94c727074b7aaac71f1560cdfc5b7d2a427d373f

Left inline comments on some of the commits.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1643,337 @@ static UniValue savemempool(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+        throw std::runtime_error(
+            "scantxoutset <action> {\"pubkeys\": [\"pubkey\",...], \"xpubs\":[{\"xpub\": \"<xpub>\"}]}\n"
```
Maybe change this to
```
    "scantxoutset \"action\" ( option )\n"
```

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1643,337 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            "                \"lookupwindow\": [<startindex>, <stopindex>] (An array with two integers that does define the range of keys that will be deriven for the given xpubs, default is 0 to 1000)\n"
+            "                }\n"
+            "           ]\n"
+            "      \"rawsweep\": {\n             (object, optional) Optionally creates a raw sweep transaction\n"
```
Remove `\n`.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1643,337 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            "           [\n"
+            "               {\n"
+            "                \"xpub\":\"<xpub>\",  (base58check encoded extended public key (xpub)\n"
+            "                \"lookupwindow\": [<startindex>, <stopindex>] (An array with two integers that does define the range of keys that will be deriven for the given xpubs, default is 0 to 1000)\n"
```
Needs 2 spaces between argument name and description.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1643,337 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            " \"rawsweep_tx\" : \"value\",       (string) The hex-encoded raw transaction of the optional sweep transaction\n"
+            " \"rawsweep_vsize\" : \"value\",     (numeric) virtual transaction size of the sweep transaction including signatures\n"
+            " \"rawsweep_fee\" : \"value\",       (numeric) Estimated fee for the sweep transaction in " + CURRENCY_UNIT + "\n"
+            "]\n"
```
Replace `]` with `}`.

promag replied:
Commit "Commit "Blockchain/RPC: Add scantxoutset method to scan UTXO set"

The response is an object, not an array.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1643,337 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            "   }\n"
+            "   ,...], \n"
+            " \"total_amount\" : x.xxx,          (numeric) The total amount of all found unspent outputs in " + CURRENCY_UNIT + "\n"
+            " \"rawsweep_tx\" : \"value\",       (string) The hex-encoded raw transaction of the optional sweep transaction\n"
```
Needs 2 spaces before description.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1643,337 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            "   ,...], \n"
+            " \"total_amount\" : x.xxx,          (numeric) The total amount of all found unspent outputs in " + CURRENCY_UNIT + "\n"
+            " \"rawsweep_tx\" : \"value\",       (string) The hex-encoded raw transaction of the optional sweep transaction\n"
+            " \"rawsweep_vsize\" : \"value\",     (numeric) virtual transaction size of the sweep transaction including signatures\n"
```
Needs 1 space before description.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1643,337 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            " \"total_amount\" : x.xxx,          (numeric) The total amount of all found unspent outputs in " + CURRENCY_UNIT + "\n"
+            " \"rawsweep_tx\" : \"value\",       (string) The hex-encoded raw transaction of the optional sweep transaction\n"
+            " \"rawsweep_vsize\" : \"value\",     (numeric) virtual transaction size of the sweep transaction including signatures\n"
+            " \"rawsweep_fee\" : \"value\",       (numeric) Estimated fee for the sweep transaction in " + CURRENCY_UNIT + "\n"
```
Needs 1 space before description.

promag replied:
Remove `,` as this is the last key.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1637,6 +1640,249 @@ static UniValue savemempool(const JSONRPCRequest& request)
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid address");
+                }
+                CScript script = GetScriptForDestination(dest);
+                if (!script.empty()) {
```
How about `assert(!script.empty())`?

promag reviewed (commented):
Some comments.

jonasschnelli commented on src/coins.cpp:
```diff
@@ -19,6 +23,41 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
     return GetCoin(outpoint, coin);
 }
 
+bool CCoinsView::FindScriptPubKey(std::atomic<int>& scanProgress, std::atomic<bool>& shouldAbort, int64_t& count, CCoinsViewCursor& cursor, const std::set<CScript>& needles, std::map<COutPoint, Coin>& out_results) {
```
Since we access the instance cursor (`CCoinsView()::Cursor()`), wouldn't it then require to access `pcoinsdbview` from within `coins.cpp` which seems not ideal?

achow101 commented on src/coins.cpp:
```diff
@@ -19,6 +23,41 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
+    return true;
+}
+
+bool CCoinsView::FindScriptPubKey(std::atomic<int>& scan_progress, std::atomic<bool>& should_abort, int64_t& searchItems, const std::set<CScript>& needles, std::map<COutPoint, Coin>& out_results) {
```
nit: `searchItems` should be `search_items` to match the `.h` file.

achow101 commented on src/rpc/blockchain.cpp:
```diff
@@ -1926,6 +1929,251 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            "               {\n"
+            "                \"xpub\":\"<xpub>\",  (base58check encoded extended public key (xpub)\n"
+            "                \"lookupwindow\":   [<startindex>, <stopindex>] (An array with two integers that does define the range of keys that will be deriven for the given xpubs, default is 0 to 1000)\n"
+            "                }\n"
```
nit: should have `,...` after this.

achow101 commented on src/rpc/blockchain.cpp:
```diff
@@ -1926,6 +1929,251 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            "    \"amount\" : x.xxx,             (numeric) The total amount in " + CURRENCY_UNIT + " received by the address\n"
+            "    \"height\" : n,                 (numeric) Height of the unspent transaction output\n"
+            "   }\n"
+            "   ,...], \n"
```
nit: the spacing of this block looks a bit funny. I think the text should be indented more and the braces for the object should be aligned.

achow101 reviewed (commented):
utACK 1945f5fd3f14d7d485fc00444cc0e8c4bdc98220

promag reviewed (commented):
Nit, remove fee handling? The caller could do:
```
tx = scantxoutset("start", { "sweep_to": address })["sweep_tx"]
fundrawtransaction(tx, { "subtractFeeFromOutputs": [0] })
...
```
So `scantxoutset` should just add the unspents and one output.

MarcoFalke commented:
Just noting that GitHub decides to hide half of my review comments by default and they were not addressed (mostly feedback on the tests).

jonasschnelli committed:
scantxoutset: add support for scripts

jonasschnelli commented:
Followed @sipa advice and changed the API.
The script type derivation is now controllable via the API.

The command takes now <n> scan objects that are either an `address`, `script`, `pubkey` or `xpub`.
Xpub and pubkey have the option to pass in the script type (P2PKH, P2SH-P2WPKH, P2WPKH).

The commit history is also more clean now.

Thanks for reviewing again.

@MarcoFalke: your review is lost in the nirvana.. can't find the comments on the test anymore. Can you try to review again?
@Sjors: the `Min relay fee not met (code -26)` is due to non initialise fee estimation. Use a feerate if you test on regtest without fee estimation available.

```
"scanobjects"                  (array, required) Array of scan objects (only one object type per scan object allowed)
      [
        { "address" : "<address>" },       (string, optional) Bitcoin address
        { "script"  : "<scriptPubKey>" },  (string, optional) HEX encoded script (scriptPubKey)
        { "pubkey"  :                      (object, optional) Public key
          {
            "pubkey" : "<pubkey">,         (string, required) HEX encoded public key
            "script_types" : [ ... ],      (array, optional) Array of script-types to derive from the pubkey (possible values: "P2PKH", "P2SH-P2WPKH", "P2WPKH")
          }
        },
        { "xpub"  :                        (object, optional) Use an extended public key child key range (m/0/k & m/1/k) to derive scripts from
          { 
            "xpub" : "<xpub">,             (string, required) Base58check encoded extended public key (xpub)
            "range" : [ <s>, <e> ],        (array, optional) Range of keys that will be deriven from the given xpubs (default is 0 to 1000)
            "script_types" : [ ... ],      (array, optional) Array of derivation type (possible values: "P2PKH", "P2SH-P2WPKH", "P2WPKH")
          }
        },
      ]
```

jonasschnelli commented:
@Sjors 
> total_amount doesn't include unconfirmed funds, maybe add a flag to show those?

This should be done in a separate command (`scanmempool` which would be a great follow up).

> a minimum block height / date for faster scans (if only to make RPC dev lives happier)

Not possible AFAIK. We scan the UXTO set, not blocks.

> it would be really cool to sweep using an xpriv (with or without importing keys)

IMO the command should not include any private keys. Ideally we add xpriv support similar to this command to signrawtransaction.
But I agree that this would be a great.

> rawsweep should have an option to make 1 transaction per output

Maybe, but could/should be added later.

> I'm assuming that calling rawsweep without the optional address field makes it use the wallet? (if so, that should be in the help)

No. This command is totally de-coupled from the wallet (runs without wallet).

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1926,6 +1932,428 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            "      ]\n"
+            "3. \"options\"                               (object, optional)\n"
+            "      \"rawsweep\": {                        (object, optional) Optionally creates a raw sweep transaction\n"
+            "          \"address\": \"address\",            (string, optional) Address where the funds should be sent to\n"
```
As @Sjors point out, this should be required if `rawsweep` is set?

jonasschnelli replied:
Fixed

promag commented:
> `rawsweep` should have an option to make 1 transaction per output

@Sjors why?

Sjors commented:
@jonasschnelli the `Min relay fee not met` error was on testnet, not regtest. Maybe that doesn't happen on mainnet much. Maybe it should detect the lack of fee estimation and throw a specific error (if no fee is specified by the user)?

> We scan the UXTO set, not blocks

I wish there an easy way to "translate" Bitcoin Core's internal data structures to SQL table descriptions, so it's easier to understand what "columns" and "indexes" exist. I haven't had much luck using external tools to load the various db files.

>> I'm assuming that calling rawsweep without the optional address field makes it use the wallet? (if so, that should be in the help)
> No. This command is totally de-coupled from the wallet (runs without wallet).

Ok, then I don't understand why the address field is documented as optional.

@promag 1 transaction per output (to a unique address) improves privacy when sweeping an existing wallet. Especially if you don't broadcast them all at the same time.

laanwj commented:
I have an unspent txout (on testnet) like this:
```json
  {
    "txid": "473abbc4eb4768d1fdad21f05d3485bcd96cb33f883dce3dd8bccebced3f4efc",
    "vout": 0,
    "address": "n2PaqLena7QZtpKowZok8FMPqCY21Xrp4x",
    "label": "null",
    "scriptPubKey": "21029b861186b49793708a4e00ecf3cec62c027dccfc12509b7e6ac2538b4b19b3abac",
    "amount": 25.00317000,
    "confirmations": 980258,
    "spendable": true,
    "solvable": true,
    "safe": true
  }
```

By address gives me no output:
```json
$ src/bitcoin-cli -testnet scantxoutset start '[{"address":"n2PaqLena7QZtpKowZok8FMPqCY21Xrp4x"}]'
{
  "success": "yes",
  "searched_items": 19143564,
  "unspents": [
  ],
  "total_amount": 0.00000000
}
$ 
```

By script it works, however
```json
$ src/bitcoin-cli -testnet scantxoutset start '[{"script":"21029b861186b49793708a4e00ecf3cec62c027dccfc12509b7e6ac2538b4b19b3abac"}]'
{
  "success": "yes",
  "searched_items": 19162149,
  "unspents": [
    {
      "txid": "473abbc4eb4768d1fdad21f05d3485bcd96cb33f883dce3dd8bccebced3f4efc",
      "vout": 0,
      "scriptPubKey": "21029b861186b49793708a4e00ecf3cec62c027dccfc12509b7e6ac2538b4b19b3abac",
      "amount": 25.00317000,
      "height": 330282
    }
  ],
  "total_amount": 25.00317000
}
```

MarcoFalke commented:
Needs rebase

jonasschnelli commented:
@laanwj:
If I execute `getaddressinfo n2PaqLena7QZtpKowZok8FMPqCY21Xrp4x` I get 
```json
{
  "address": "n2PaqLena7QZtpKowZok8FMPqCY21Xrp4x",
  "scriptPubKey": "76a914e4f5d0b9353f7715e1bfed11358a5b84fd7c4e3488ac",
  "ismine": false,
  "iswatchonly": false,
  "isscript": false,
  "iswitness": false,
  "labels": [
  ]
}
```

AFAIK your script (P2PK) `21029b861186b49793708a4e00ecf3cec62c027dccfc12509b7e6ac2538b4b19b3abac` can only be derived if one has the according pubkey.

jonasschnelli commented:
Rebased.

jonasschnelli committed:
scantxoutset: support legacy P2PK script type

jonasschnelli commented:
Added two commits:

https://github.com/bitcoin/bitcoin/pull/12196/commits/fdd09d34a14ca4f416b9101b37187098f9145301 
adds support for P2PK scripts type

https://github.com/bitcoin/bitcoin/pull/12196/commits/3c546fd38e2bdb6957fa134f8847e2a7c5c7e23

P2PK and P2PKH share the same address (two scripts result in the same address). Deriving a P2PK script requires the pubkey which is not available when the user provides an address.

Since we not want to miss funds, we have to derive the P2PKH equivalent of every P2PK unspent in the txoutset and compare it against the provided address (only required when providing addresses).

MarcoFalke commented on src/coins.h:
```diff
@@ -157,6 +160,10 @@ class CCoinsView
 
+    //! Search for a given set of pubkey scripts
+    static bool FindScriptPubKey(std::atomic<int>& scan_progress, std::atomic<bool>& should_abort, int64_t& count, CCoinsViewCursor& cursor, std::function<bool(const CScript& script)> match_func, std::map<COutPoint, Coin>& out_results);
+    bool FindScriptPubKey(std::atomic<int>& scan_progress, std::atomic<bool>& should_abort, int64_t& search_items, std::function<bool(const CScript& script)> match_func, std::map<COutPoint, Coin>& out_results);
```
`std::function` is defined in header `<functional>`, which needs to be included.

jonasschnelli committed:
scantxoutset: mention that scanning by address will miss P2PK txouts

jonasschnelli commented:
After discussion with @sipa on IRC, I removed the auto-derivation for P2PK scripts when scanning with addresses. Instead, I have added a short info to the RPC help that should make users aware for that particular edge case.

promag reviewed (commented):
Some comments.

jonasschnelli commented on src/coins.cpp:
```diff
@@ -19,6 +23,41 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
+        if (cursor.GetKey(key) && cursor.GetValue(coin)) {
+            if (count++ % 8192 == 0) {
+                boost::this_thread::interruption_point();
+                if (should_abort) {
```
This would couple it to init.h,.. ideally we would finally extract the shutdown detection into its own space.

laanwj commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,435 @@ static UniValue savemempool(const JSONRPCRequest& request)
+        throw std::runtime_error(
+            "scantxoutset <action> <scanobjects> ( <options> )\n"
+            "\nScans the unspent transaction output set for possible entries that matches common scripts of given public keys.\n"
+            "Using addresses as scanobjects will _not_ detect unspent P2PK txouts\n"
```
Ok, fair enough.

luke-jr commented on src/coins.cpp:
```diff
@@ -19,6 +23,39 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
+    while (cursor.Valid()) {
+        COutPoint key;
+        Coin coin;
+        if (!cursor.GetKey(key) || !cursor.GetValue(coin)) return false;
```
This is refactored to a short-circuit in `Blockchain/RPC: Add scantxoutset method to scan UTXO set` after being added in `Add CCoinsView::FindScriptPubKey to search the UTXO set`. Instead, it should just be added refactored initially.

luke-jr commented:
Again, addresses are opaque and do not get tied to UTXOs. It makes no sense to find UTXOs "by address".

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,435 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            "    \"txid\" : \"transactionid\",     (string) The transaction id\n"
+            "    \"vout\": n,                    (numeric) the vout value\n"
+            "    \"scriptPubKey\" : \"script\",    (string) the script key\n"
+            "    \"amount\" : x.xxx,             (numeric) The total amount in " + CURRENCY_UNIT + " received by the address\n"
```
s/received by the address/of the UTXO/

jonasschnelli replied:
Fixed.

jonasschnelli commented:
@luke-jr 
> Again, addresses are opaque and do not get tied to UTXOs. It makes no sense to find UTXOs "by address".

You have already made that argument [here](https://github.com/bitcoin/bitcoin/pull/12196#issuecomment-360244898).
It seems like that the scans based on addresses are useful. Other developers opinions would be welcome.

laanwj commented:
> It seems like that the scans based on addresses are useful. Other developers opinions would be welcome.

Yes, that is useful, and for better or worse, likely how most people will be using this in practice.

ajtowns commented:
> Addresses are opaque identifiers for a given invoice. That they are currently implemented by encoding a scriptPubKey is irrelevant.

This makes no sense to me: per the wiki "A Bitcoin address [...] is an identifier [...] that represents a possible destination for a bitcoin payment", but if you want to send a bitcoin payment somewhere you have to be able to deduce the scriptPubKey from the address, or you can't make a transaction. I don't think it makes sense to generalise "bitcoin addresses" to things that you can't generate a scriptPubKey from (eg, bech32 encoded lightning invoices, or some new address format you don't yet understand).

Finding UTXOs by scriptPubKey seems like it makes sense, and per the above finding the scriptPubKeys from an address should also always be straightforward. So ConceptACK fwiw, unless I'm missing something major here...

laanwj commented:
Fails to build locally here:
```
/.../bitcoin/src/bench/block_assemble.cpp:42:13: error: use of undeclared identifier 'CheckProofOfWork'
    while (!CheckProofOfWork(block->GetHash(), block->nBits, Params().GetConsensus())) {
            ^
```

promag commented on test/functional/rpc_scantxoutset.py:
```diff
@@ -0,0 +1,76 @@
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the scantxoutset rpc call."""
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import *
```
nit, could specify symbols instead.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,435 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            // make sure only one object type is present
+            if (1 != !address_uni.isNull() + !pubkey_uni.isNull() + !script_uni.isNull() + !xpub_uni.isNull()) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Only one object type is allowed per scan object");
+            }
```
nit, `} else if (...) {`

There are a couple of more cases throughout.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,435 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            total_in += txo.nValue;
+
+            UniValue unspent(UniValue::VOBJ);
+            unspent.push_back(Pair("txid", outpoint.hash.GetHex()));
```
Should use `pushKV` instead, `push_back(Pair(...))` occurrences were replaced recently.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,435 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            const CScript& scriptPubKey = txo.scriptPubKey;
+            SignatureData sigdata;
+            ProduceSignature(temp_keystore, DUMMY_SIGNATURE_CREATOR, scriptPubKey, sigdata);
+            UpdateInput(tx.vin.at(nIn), sigdata);
```
Could drop `nIn` and use `tx.vin.back()` instead (like above).

jonasschnelli replied:
More readable like this IMO

promag reviewed (commented):
Some nits and comments. Tests LGTM. Could squash last commit.

utACK.

Empact commented on src/coins.cpp:
```diff
@@ -19,6 +23,39 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
+        COutPoint key;
+        Coin coin;
+        if (!cursor.GetKey(key) || !cursor.GetValue(coin)) return false;
+        if (count++ % 8192 == 0) {
```
nit: `++count`

jonasschnelli replied:
Fixed

Empact commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,435 @@ static UniValue savemempool(const JSONRPCRequest& request)
+    explicit CoinsViewScanReserver() : m_could_reserve(false) {}
+
+    bool reserve() {
+        assert (!m_could_reserve);
```
`include <assert.h>`
nit: whitespace could benefit from clang-format

jonasschnelli replied:
Fixed

promag replied:
Commit "Blockchain/RPC: Add scantxoutset method to scan UTXO set"

Missing fix :trollface:

Empact commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,435 @@ static UniValue savemempool(const JSONRPCRequest& request)
+    else if (outputtype == "P2PKH") return OutputScriptType::P2PKH;
+    else if (outputtype == "P2SH_P2WPKH") return OutputScriptType::P2SH_P2WPKH;
+    else if (outputtype == "P2WPKH") return OutputScriptType::P2WPKH;
+    else return OutputScriptType::UNKNOWN;
```
nit: this is basically an ideal case for a `case` statement. :P

jimpo replied:
You can't switch on strings in C/C++

Empact replied:
Ah, thanks.

Empact commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,435 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            }
+            else if (!script_uni.isNull() && !script_uni.isStr()) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Scanobject \"xpub\" must contain a single string as value");
+            }
```
You could consider using `RPCTypeCheckObj` for some of this. E.g. https://github.com/bitcoin/bitcoin/blob/686e97a0c7358291d628213447cf33e99cde7ce8/src/rpc/rawtransaction.cpp#L786-L791

sipa commented:
As discussed in the meeting (http://www.erisian.com.au/meetbot/bitcoin-core-dev/2018/bitcoin-core-dev.2018-06-28-19.00.log.html), my suggestion is to (for now) drop the xpub support from this PR. That way we have some more time for a generic notation for chains of scripts to develop (something I'm currently working on) and be compatible with other places such a notation may be adopted.

laanwj commented:
Agree with @sipa - that way, we can move this forward without full agreement on xpub representation.

jonasschnelli commented:
Removed the xpub support. Thanks for reviewing again.

laanwj commented:
re-utACK 3bf8dbe48a1f45b7aa0584bfb0cf19a7ddf492c9

laanwj commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1923,356 @@ static UniValue savemempool(const JSONRPCRequest& request)
+                    if (script_type == OutputScriptType::P2PK) {
+                        // support legacy P2PK scripts
+                        script << ToByteVector(pubkey) << OP_CHECKSIG;
+                    }
```
nit: `} else {`

jonasschnelli replied:
Fixed

jonasschnelli commented on src/coins.cpp:
```diff
@@ -19,6 +23,41 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
     return GetCoin(outpoint, coin);
 }
 
+bool CCoinsView::FindScriptPubKey(std::atomic<int>& scan_progress, std::atomic<bool>& should_abort, int64_t& count, CCoinsViewCursor& cursor, const std::set<CScript>& needles, std::map<COutPoint, Coin>& out_results) {
```
Fixed.

achow101 replied:
The change is missing.

jonasschnelli commented on src/coins.cpp:
```diff
@@ -19,6 +23,41 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
 }
 
+bool CCoinsView::FindScriptPubKey(std::atomic<int>& scan_progress, std::atomic<bool>& should_abort, int64_t& count, CCoinsViewCursor& cursor, const std::set<CScript>& needles, std::map<COutPoint, Coin>& out_results) {
+    scan_progress = 0;
```
Fixed.

jonasschnelli commented on src/coins.cpp:
```diff
@@ -19,6 +23,41 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
+    while (cursor.Valid()) {
+        COutPoint key;
+        Coin coin;
+        if (cursor.GetKey(key) && cursor.GetValue(coin)) {
```
Fixed.

jonasschnelli reviewed (commented):
Fixed nits.

promag commented:
While testing I hit an edge case:
```
./src/bitcoin-cli -regtest  scantxoutset start '[]' & ;  ./src/bitcoin-cli -regtest  scantxoutset abort
[1] 90388
true
{
  "success": "yes",
  "searched_items": 6,
  "unspents": [
  ],
  "total_amount": 0.00000000
}
[1]  + 90388 done       ./src/bitcoin-cli -regtest scantxoutset start '[]'
```
So `abort` returned `true` but the scan finished successfully.

Not sure but if the use case is to call `abort` after `status` — because it's taking too long — so maybe `abort` could just return null:
```cpp
if (request.params[0].get_str() == "abort") {
    g_should_abort_scan = true;
    return NullUniValue;
}
```

laanwj commented:
I'm not sure that's a harmful edge case. If you abort after the last opportunity to abort, I'd say it makes sense that it will finish successfully.

promag commented:
That's why I ask if it makes sense to return a bool in abort.

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -2012,6 +2032,12 @@ UniValue scantxoutset(const JSONRPCRequest& request)
             "        },\n"
             "      ]\n"
+            "3. \"options\"                               (object, optional)\n"
+            "      \"rawsweep\": {                        (object, optional) Optionally creates a raw sweep transaction\n"
```
In commit "scantxoutset: Add optional raw sweep transaction".

Is this the best place to implement this? As a separate wallet RPC it could work much more correctly (as it would know the actual scripts involved and be able to give a vsize estimate that's more than a dumb guess), and be more useful too by being combinable with whatever way to find inputs.

I feel this makes the RPC look like a kitchen sink.

jonasschnelli replied:
Agree. Will remove it and PR it later as a separate method

promag replied:
From https://github.com/bitcoin/bitcoin/pull/12196#issuecomment-393057791

> > tx = scantxoutset("start", { "sweep_to": address })["sweep_tx"]
> > fundrawtransaction(tx, { "subtractFeeFromOutputs": [0] })
>
>  @promag: would that also work without the wallet? My idea was to make `scantxoutset` work without a wallet.

@sipa not sure if you are suggesting something similar to:
```
unspents = scantxoutset("start", ...)
tx = createrawtransaction unspents ...
tx = fundrawtransaction ...
```
And how could it work without the wallet?

sipa replied:
@promag, I just mean scantxoutset + createrawtransaction. Then a separate wallet RPC like "walletestimatesignedrawtransactionsize" (shorter names welcome...) can be used to guess the size of the tx ahead of time. "fundrawtransaction" already does this estimation internally.

> Would that also work without the wallet?

No, that's the point. You cannot implement that without knowledge of the scripts involved.

It may be better to integrate this into PSBT, as it means there doesn't need to a single wallet that knows about all scripts in a transaction.

laanwj commented on src/coins.cpp:
```diff
@@ -19,6 +23,41 @@ bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
     return GetCoin(outpoint, coin);
 }
 
+bool CCoinsView::FindScriptPubKey(std::atomic<int>& scan_progress, std::atomic<bool>& should_abort, int64_t& count, CCoinsViewCursor& cursor, const std::set<CScript>& needles, std::map<COutPoint, Coin>& out_results) {
```
Agree—if we can avoid introducing new methods to the CCoinsView interface, that's preferable, it is good to have it as a minimal interface class.

jonasschnelli commented:
* Moved the `FindScriptPubKey()` method to `rpc/blockchain.cpp` (changes are no longer spread in multiple commits).
* Removed the sweep transaction creation code (see https://github.com/bitcoin/bitcoin/pull/12196#pullrequestreview-134880350)

I kept the pubkey search option since I think this is very helpful. I agree that we should move to the descriptors specified by @sipa. IMO we define that we wanted to have xpub/pub support (according those descriptors) in the same release to avoid API breaks and @laanwj agreed to let this in after the code freeze in form of a bugfix.
Additionally, there is always the option to mark this RPC command "experimental" (API changes possible).

luke-jr reviewed (changes requested):
Maintaining a NACK so long as it confuses addresses (which are unrelated to UTXOs) with keys...

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1945,6 +1950,246 @@ bool FindScriptPubKey(std::atomic<int>& scan_progress, const std::atomic<bool>&
+    case OutputScriptType::P2WPKH: {
+        if (!key.IsCompressed()) return key.GetID();
+        CTxDestination witdest = WitnessV0KeyHash(key.GetID());
+        CScript witprog = GetScriptForDestination(witdest);
```
Commit "Blockchain/RPC: Add scantxoutset method to scan UTXO set"

This could be in the block below:
```cpp
CTxDestination witdest = WitnessV0KeyHash(key.GetID());
if (type == OutputScriptType::P2SH_P2WPKH) {
    CScript witprog = GetScriptForDestination(witdest);
    return CScriptID(witprog);
} else {
    ...
```

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1945,6 +1950,246 @@ bool FindScriptPubKey(std::atomic<int>& scan_progress, const std::atomic<bool>&
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+        throw std::runtime_error(
+            "scantxoutset <action> <scanobjects> ( <options> )\n"
```
Commit "Blockchain/RPC: Add scantxoutset method to scan UTXO set"

nit, add `( )` to `<scanobjects>` because `status` action has no more arguments.

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -1945,6 +1950,246 @@ bool FindScriptPubKey(std::atomic<int>& scan_progress, const std::atomic<bool>&
+        int64_t count = 0;
+        FlushStateToDisk();
+
+        std::unique_ptr<CCoinsViewCursor> pcursor(pcoinsdbview->Cursor());
```
You should hold cs_main while calling FlushStateToDisk and creating the cursor, or there is no guarantee that the on-disk state is consistent.

promag replied:
Looks like some annotations or lock assertions could be added?

jonasschnelli replied:
Agree with @promag, but unrelated to this PR.

promag replied:
Oh, I wasn't implying that. I'll follow up.

promag replied:
Ops, actually @sipa point is about locking during `FlushStateToDisk()` and `Cursor()`, so the lock must be added here, preferably in this PR IMO.

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -1945,6 +1950,246 @@ bool FindScriptPubKey(std::atomic<int>& scan_progress, const std::atomic<bool>&
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Scan already in progress, use action \"abort\" or \"status\"");
+        }
+        std::set<CScript> needles;
+        CBasicKeyStore temp_keystore;
```
This looks unused.

jonasschnelli commented:
Fixed issues:
* holds now `cs_main` during coins view cursor creation and `FlushToDisk()`
* Removed `temp_keystore` (relict from raw-sweep-tx).
* Fixed "yes"/"no" instead of true/false JSON returns
* Fixed nits

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,289 @@ static UniValue savemempool(const JSONRPCRequest& request)
+        if (count % 256 == 0) {
+            // update progress reference every 256 item
+            uint32_t high = 0x100 * *key.hash.begin() + *(key.hash.begin() + 1);
+            scan_progress = (int)(high * 100.0 / 65536.0 + 0.5);
```
Nit: code for this seems like it's cropping up in a few different places, might be nice to have an abstraction for it.
```
 $ git grep -C 1 "/ 65536" | cat

src/index/txindex.cpp-                (static_cast<uint32_t>(*(txid.begin() + 1)) << 0);
src/index/txindex.cpp:            int percentage_done = (int)(high_nibble * 100.0 / 65536.0 + 0.5);
src/index/txindex.cpp-
--
src/txdb.cpp-                uint32_t high = 0x100 * *key.second.begin() + *(key.second.begin() + 1);
src/txdb.cpp:                int percentageDone = (int)(high * 100.0 / 65536.0 + 0.5);
src/txdb.cpp-                uiInterface.ShowProgress(_("Upgrading UTXO database"), percentageDone, true);
```

jonasschnelli replied:
IMO this should be done outside of this PR

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,289 @@ static UniValue savemempool(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+        throw std::runtime_error(
+            "scantxoutset <action> ( <scanobjects> ) ( <options> )\n"
```
If scanobjects is required (per doc below), should it be in parens here?

jamesob replied:
Also options doesn't seem to exist.

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,289 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            FlushStateToDisk();
+            std::unique_ptr<CCoinsViewCursor> pcursor(pcoinsdbview->Cursor());
+            assert(pcursor);
+            bool res = FindScriptPubKey(g_scan_progress, g_should_abort_scan, count, pcursor.get(), needles, coins);
```
Can we release `cs_main` after initializing `pcursor`? My impression is that CCoinsViewCursor can be used for read-only ops without holding it.

jonasschnelli replied:
Oh. Right. Fixed.

jamesob commented:
Concept ACK

Happy to do a more in-depth review. PR description looks in need of an update.

jonasschnelli commented:
Fixed points reported by @jamesob.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,290 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            assert(pcursor);
+        }
+        bool res = FindScriptPubKey(g_scan_progress, g_should_abort_scan, count, pcursor.get(), needles, coins);
+        result.pushKV("success", res ? true : false);
```
`res` is already a boolean...

jonasschnelli replied:
Ouch! Fixed.

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -2141,8 +2141,13 @@ UniValue scantxoutset(const JSONRPCRequest& request)
+                    CScript script;
+                    if (script_type == OutputScriptType::P2PK) {
+                        // support legacy P2PK scripts
+                        script << ToByteVector(pubkey) << OP_CHECKSIG;
```
Nit: use `GetScriptForRawPubKey`.

sipa reviewed (commented):
utACK b9b59cc7ba742bb519d34f3011500f09a8c04911

Empact commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,290 @@ static UniValue savemempool(const JSONRPCRequest& request)
+bool FindScriptPubKey(std::atomic<int>& scan_progress, const std::atomic<bool>& should_abort, int64_t& count, CCoinsViewCursor* cursor, const std::set<CScript>& needles, std::map<COutPoint, Coin>& out_results) {
+    scan_progress = 0;
+    count = 0;
+    while (cursor->Valid()) {
```
nit: `for` would localize the `cursor` access

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,290 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            } else if (!pubkey_uni.isNull() && !pubkey_uni.isObject()) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Scanobject \"pubkey\" must contain an object as value");
+            } else if (!script_uni.isNull() && !script_uni.isStr()) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Scanobject \"script\" must contain a single string as value");
```
I guess your proposed check does not ensure that one of the (either address, pubkey or script must be present, right?

Empact replied:
Right, it just handles the type testing.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,290 @@ static UniValue savemempool(const JSONRPCRequest& request)
+            UniValue script_uni  = find_value(scanobject, "script");
+
+            // make sure only one object type is present
+            if (1 != !address_uni.isNull() + !pubkey_uni.isNull() + !script_uni.isNull()) {
```
I think it helps for the readability? Or can you elaborate what you mean exactly?

Empact replied:
Suggesting to group the sum in parens so that it's more clear that there is only the one top-level condition.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,290 @@ static UniValue savemempool(const JSONRPCRequest& request)
+                // loop through the script types and derive the script
+                for (const UniValue& script_type_uni : script_types_uni.get_array().getValues()) {
+                    OutputScriptType script_type = GetOutputScriptTypeFromString(script_type_uni.get_str());
+                    if (script_type == OutputScriptType::UNKNOWN) throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid script type");
```
I don't understand your comment... can you rephrase or elaborate in detail?

Empact replied:
If you `throw` on `UNKNOWN` within `GetOutputScriptTypeFromString`, then there is no other use of `UNKNOWN` as an `OutputScriptType`.

jonasschnelli commented on test/functional/rpc_scantxoutset.py:
```diff
@@ -0,0 +1,48 @@
+        self.log.info("Test if we have found the non HD unspent outputs.")
+        assert_equal(self.nodes[0].scantxoutset("start", [ {"pubkey": {"pubkey": pubk1}}, {"pubkey": {"pubkey": pubk2}}, {"pubkey": {"pubkey": pubk3}}])['total_amount'], 6)
+        assert_equal(self.nodes[0].scantxoutset("start", [ {"address": addr_P2SH_SEGWIT}, {"address": addr_LEGACY}, {"address": addr_BECH32}])['total_amount'], 6)
+        assert_equal(self.nodes[0].scantxoutset("start", [ {"address": addr_P2SH_SEGWIT}, {"address": addr_LEGACY}, {"pubkey": {"pubkey": pubk3}} ])['total_amount'], 6)
```
Both would probably require additional code for adding time delays or temporary creation of a large utxo set.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,290 @@ static UniValue savemempool(const JSONRPCRequest& request)
+    }
+};
+
+const char *g_default_scantxoutset_script_types[] = { "P2PKH", "P2SH_P2WPKH", "P2WPKH" };
```
Agree. Made static.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,290 @@ static UniValue savemempool(const JSONRPCRequest& request)
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Public key must be hex encoded");
+                }
+                std::vector<unsigned char> data(ParseHexV(pubkeydata_uni, "pubkey"));
+                CPubKey pubkey(data.begin(), data.end());
```
Indeed. Fixed.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -1916,6 +1920,290 @@ static UniValue savemempool(const JSONRPCRequest& request)
+                // type: script
+                // check and add the script to the scan containers (needles array)
+                std::vector<unsigned char> scriptData(ParseHexV(script_uni, "script"));
+                CScript script(scriptData.begin(), scriptData.end());
```
Fixed.

jonasschnelli commented:
Fixed relevant points from @Empact

promag commented:
utACK be98b2d.

surindergiri commented:
Thanks for your great efforts implementing this feature, On Oct 01, 2020, When I run this RPC command on Bitcoin Node version 0.20.0, then its response is too slow, I am getting rpc response after 1-2 minutes. 

Is there any way to speed up this "scantxoutset" RPC command or any alternative available?

I asked the same question over bitcoin stackexchange forum, but nobody responded. I know its not the right place to ask such questions, but I am stuck, our mobile wallet is live and we are unable to fetch user "imported" address balances on the fly.

Thanks.

https://bitcoin.stackexchange.com/questions/99278/bitcoin-scantxoutset-rpc-command-return-response-too-slow

jonatack commented:
@surindergiri please see/test/review #18000 and its separate PRs, which will make this rpc run much more quickly.
